File: backend/app/utils/geometry.py
import math
import numpy as np

FEATURE_NAMES = [
    "elbow_l", "elbow_r", "shoulder_l", "shoulder_r",
    "hip_l", "hip_r", "knee_l", "knee_r",
    "ankle_l", "ankle_r", "trunk_l", "trunk_r",
    "neck", "hip_abduct_l", "hip_abduct_r"
]

def calculate_angle_3d(a, b, c):
    """Calculates the 3D angle between vector BA and vector BC. Point B is the vertex."""
    ba = a - b
    bc = c - b
    
    dot_product = np.dot(ba, bc)
    norm_ba = np.linalg.norm(ba)
    norm_bc = np.linalg.norm(bc)
    
    if norm_ba == 0 or norm_bc == 0:
        return 180.0
        
    cosine_angle = dot_product / (norm_ba * norm_bc)
    cosine_angle = np.clip(cosine_angle, -1.0, 1.0)
    
    angle = np.arccos(cosine_angle)
    return float(np.degrees(angle))

SHOULDER_L, SHOULDER_R = 11, 12
ELBOW_L, ELBOW_R = 13, 14
WRIST_L, WRIST_R = 15, 16
HIP_L, HIP_R = 23, 24
KNEE_L, KNEE_R = 25, 26
ANKLE_L, ANKLE_R = 27, 28
HEEL_L, HEEL_R = 29, 30
NOSE = 0

def extract_angles_from_landmarks(points: np.ndarray, zero_z: bool = True) -> list:
    """
    Mirrors frontend/src/utils/geometry.ts's extractAnglesFromLandmarks, so the
    Gradio demo (which runs MediaPipe server-side on an uploaded/webcam image)
    computes the exact same 15 biomechanical features, in the same order as
    FEATURE_NAMES, that the client-side pipeline sends to /api/analyse_frame.
    points shape: [33, 3] (x, y, z per MediaPipe landmark)

    MediaPipe's default (image-normalized) z-depth estimate is only reliable
    at the consistent camera distance/framing seen in demo videos; on
    arbitrary real-world camera framing it degrades badly (measured 0-3%
    real-world pose accuracy with z included, vs ~46% with it dropped,
    confirmed across a 40-trial randomized threshold sweep). zero_z=True (the
    default) reproduces that proven 2D-only path. zero_z=False is for the
    separately-calibrated `pose_world_landmarks` (metric-scale 3D) signal,
    which keeps genuine depth information and is used as an independent
    second vote alongside the 2D path rather than as a replacement for it
    (see rules_classifier.hybrid_classify).
    """
    if points.shape[0] < 31:
        return [0.0] * 15

    points = points.copy()
    if zero_z:
        points[:, 2] = 0.0

    shoulder_mid = (points[SHOULDER_L] + points[SHOULDER_R]) / 2.0
    hip_mid = (points[HIP_L] + points[HIP_R]) / 2.0

    return [
        calculate_angle_3d(points[SHOULDER_L], points[ELBOW_L], points[WRIST_L]),
        calculate_angle_3d(points[SHOULDER_R], points[ELBOW_R], points[WRIST_R]),
        calculate_angle_3d(points[HIP_L], points[SHOULDER_L], points[ELBOW_L]),
        calculate_angle_3d(points[HIP_R], points[SHOULDER_R], points[ELBOW_R]),
        calculate_angle_3d(points[SHOULDER_L], points[HIP_L], points[KNEE_L]),
        calculate_angle_3d(points[SHOULDER_R], points[HIP_R], points[KNEE_R]),
        calculate_angle_3d(points[HIP_L], points[KNEE_L], points[ANKLE_L]),
        calculate_angle_3d(points[HIP_R], points[KNEE_R], points[ANKLE_R]),
        calculate_angle_3d(points[KNEE_L], points[ANKLE_L], points[HEEL_L]),
        calculate_angle_3d(points[KNEE_R], points[ANKLE_R], points[HEEL_R]),
        calculate_angle_3d(points[SHOULDER_L], points[HIP_L], points[HIP_R]),
        calculate_angle_3d(points[SHOULDER_R], points[HIP_R], points[HIP_L]),
        calculate_angle_3d(points[NOSE], shoulder_mid, hip_mid),
        calculate_angle_3d(points[HIP_R], points[HIP_L], points[KNEE_L]),
        calculate_angle_3d(points[HIP_L], points[HIP_R], points[KNEE_R]),
    ]

def normalize_coordinate_sequence(coords: np.ndarray) -> np.ndarray:
    """
    Translates joints to be pelvis-centered (midpoint of left and right hips)
    and scales by hip-width to ensure translation and scale invariance.
    coords shape: [60, 99]
    """
    coords_reshaped = coords.reshape(60, 33, 3)
    hip_l = coords_reshaped[:, 23, :]
    hip_r = coords_reshaped[:, 24, :]
    pelvis = (hip_l + hip_r) / 2.0
    
    # Translate
    coords_normalized = coords_reshaped - pelvis[:, None, :]
    
    # Scale by hip width
    hip_width = np.linalg.norm(hip_l - hip_r, axis=-1, keepdims=True)
    hip_width = np.where(hip_width < 1e-5, 1.0, hip_width)
    coords_normalized = coords_normalized / hip_width[:, None, :]
    
    return coords_normalized.reshape(60, 99)


# --- Global body orientation -------------------------------------------------
#
# The 15 features in FEATURE_NAMES are all RELATIVE joint angles
# (shoulder-hip-knee and friends), which makes them invariant to rotating the
# whole body. Measured consequence on the real-photo corpus: a person lying
# flat in corpse and a person standing in mountain_pose produce nearly the same
# 15-vector, and the rule engine called corpse "mountain_pose" 11 times out of
# 18. No threshold change can fix that -- the information is simply absent from
# the feature vector.
#
# These two scalars restore it. Both are translation- and scale-invariant and
# come from landmarks the client already has, so they cost nothing to compute
# and nothing extra to transmit beyond two floats.

def compute_orientation(points: np.ndarray) -> dict:
    """Global orientation cues that the relative joint angles cannot express.

    torso_incline: degrees between the shoulders->hips vector and image-down.
        ~0 standing, ~90 lying flat or in plank, ~180 inverted (downward dog).
        Measured medians: mountain 6, cobra 43, table_top 61, triangle 71,
        plank 80, corpse 107, downward_dog 148.

    leg_torso_ratio: mean hip->ankle distance over shoulder->hip distance, in
        image space. Cleanly separates cross-legged sitting (0.57) from every
        standing pose (~1.2-1.4), which bounding-box aspect ratio does NOT do
        reliably (a standing person with arms down is just as tall and narrow
        as a seated one photographed head-on).

    Returns {} when the landmarks are degenerate, so callers can treat missing
    orientation the same as an older client that never sends it.
    """
    p = np.asarray(points, dtype=np.float64)[:, :2]
    shoulder_mid = (p[SHOULDER_L] + p[SHOULDER_R]) / 2.0
    hip_mid = (p[HIP_L] + p[HIP_R]) / 2.0

    torso_vec = hip_mid - shoulder_mid
    torso_len = float(np.linalg.norm(torso_vec))
    if torso_len < 1e-9:
        return {}

    # MediaPipe's y axis grows downward, so image-down is (0, +1) and the
    # inclination is just the angle of the torso vector against it.
    cos_i = float(torso_vec[1] / torso_len)
    torso_incline = math.degrees(math.acos(max(-1.0, min(1.0, cos_i))))

    leg_len = (float(np.linalg.norm(p[ANKLE_L] - p[HIP_L])) +
               float(np.linalg.norm(p[ANKLE_R] - p[HIP_R]))) / 2.0

    return {
        "torso_incline": torso_incline,
        "leg_torso_ratio": leg_len / torso_len,
    }

File: frontend/src/utils/geometry.ts
export interface Point3D {
  x: number;
  y: number;
  z: number;
}

export function calculateAngle3D(a: Point3D, b: Point3D, c: Point3D): number {
  const ba = { x: a.x - b.x, y: a.y - b.y, z: a.z - b.z };
  const bc = { x: c.x - b.x, y: c.y - b.y, z: c.z - b.z };
  
  const dot = ba.x * bc.x + ba.y * bc.y + ba.z * bc.z;
  const magBA = Math.sqrt(ba.x * ba.x + ba.y * ba.y + ba.z * ba.z);
  const magBC = Math.sqrt(bc.x * bc.x + bc.y * bc.y + bc.z * bc.z);
  
  if (magBA === 0 || magBC === 0) return 0;
  
  let cosTheta = dot / (magBA * magBC);
  cosTheta = Math.max(-1, Math.min(1, cosTheta));
  
  return (Math.acos(cosTheta) * 180.0) / Math.PI;
}

// MediaPipe landmarks indices
export const SHOULDER_L = 11;
export const SHOULDER_R = 12;
export const ELBOW_L = 13;
export const ELBOW_R = 14;
export const WRIST_L = 15;
export const WRIST_R = 16;
export const HIP_L = 23;
export const HIP_R = 24;
export const KNEE_L = 25;
export const KNEE_R = 26;
export const ANKLE_L = 27;
export const ANKLE_R = 28;
export const HEEL_L = 29;
export const HEEL_R = 30;
export const NOSE = 0;

export function extractAnglesFromLandmarks(rawPts: Point3D[], zeroZ: boolean = true): number[] {
  if (rawPts.length < 31) return Array(15).fill(0);

  // MediaPipe's monocular z-depth estimate (on the default poseLandmarks) is
  // only reliable at the consistent camera distance/framing seen in demo
  // videos; on arbitrary real-world camera framing it degrades badly
  // (measured 0-3% real-world pose accuracy with z included, vs ~46% with it
  // dropped, confirmed across a 40-trial randomized threshold sweep). zeroZ
  // defaults to true to reproduce that proven 2D-only path. Pass zeroZ=false
  // for poseWorldLandmarks (metric-scale, separately calibrated 3D) so its
  // genuine depth is kept -- used as an independent second vote server-side,
  // not a replacement for the 2D path.
  const pts = rawPts.map((p) => ({ x: p.x, y: p.y, z: zeroZ ? 0 : p.z }));

  const shoulder_mid = {
    x: (pts[SHOULDER_L].x + pts[SHOULDER_R].x) / 2.0,
    y: (pts[SHOULDER_L].y + pts[SHOULDER_R].y) / 2.0,
    z: (pts[SHOULDER_L].z + pts[SHOULDER_R].z) / 2.0,
  };
  
  const hip_mid = {
    x: (pts[HIP_L].x + pts[HIP_R].x) / 2.0,
    y: (pts[HIP_L].y + pts[HIP_R].y) / 2.0,
    z: (pts[HIP_L].z + pts[HIP_R].z) / 2.0,
  };
  
  return [
    calculateAngle3D(pts[SHOULDER_L], pts[ELBOW_L], pts[WRIST_L]), // elbow_l
    calculateAngle3D(pts[SHOULDER_R], pts[ELBOW_R], pts[WRIST_R]), // elbow_r
    calculateAngle3D(pts[HIP_L], pts[SHOULDER_L], pts[ELBOW_L]), // shoulder_l
    calculateAngle3D(pts[HIP_R], pts[SHOULDER_R], pts[ELBOW_R]), // shoulder_r
    calculateAngle3D(pts[SHOULDER_L], pts[HIP_L], pts[KNEE_L]), // hip_l
    calculateAngle3D(pts[SHOULDER_R], pts[HIP_R], pts[KNEE_R]), // hip_r
    calculateAngle3D(pts[HIP_L], pts[KNEE_L], pts[ANKLE_L]), // knee_l
    calculateAngle3D(pts[HIP_R], pts[KNEE_R], pts[ANKLE_R]), // knee_r
    calculateAngle3D(pts[KNEE_L], pts[ANKLE_L], pts[HEEL_L]), // ankle_l
    calculateAngle3D(pts[KNEE_R], pts[ANKLE_R], pts[HEEL_R]), // ankle_r
    calculateAngle3D(pts[SHOULDER_L], pts[HIP_L], pts[HIP_R]), // trunk_l
    calculateAngle3D(pts[SHOULDER_R], pts[HIP_R], pts[HIP_L]), // trunk_r
    calculateAngle3D(pts[NOSE], shoulder_mid, hip_mid), // neck
    calculateAngle3D(pts[HIP_R], pts[HIP_L], pts[KNEE_L]), // hip_abduct_l
    calculateAngle3D(pts[HIP_L], pts[HIP_R], pts[KNEE_R]), // hip_abduct_r
  ];
}

/**
 * Global body orientation, which the 15 angle features cannot express.
 *
 * Every entry in FEATURE_NAMES is a RELATIVE joint angle (shoulder-hip-knee
 * and friends), so all of them are invariant to rotating the whole body.
 * Measured on the real-photo corpus, that blindness is not theoretical: the
 * rule engine called corpse "mountain_pose" 11 times out of 18, because a
 * person lying flat and a person standing upright produce nearly the same
 * 15-vector. No threshold change can fix it -- the information is simply not
 * in the feature vector.
 *
 * torsoIncline    degrees between shoulders->hips and image-down.
 *                 ~0 standing, ~90 lying or plank, ~180 inverted.
 *                 Measured medians: mountain 6, cobra 43, table_top 61,
 *                 triangle 71, plank 80, corpse 107, downward_dog 148.
 * legTorsoRatio   mean hip->ankle distance over shoulder->hip distance.
 *                 Separates cross-legged sitting (0.57) from every standing
 *                 pose (~1.2-1.4). Bounding-box aspect ratio does NOT do this
 *                 reliably: a standing person with arms down is just as tall
 *                 and narrow as someone seated facing the camera.
 *
 * Returns null on degenerate landmarks so the caller can simply omit the
 * field, which the backend treats exactly like an older client.
 */
export function computeOrientation(
  landmarks: { x: number; y: number }[],
): { torso_incline: number; leg_torso_ratio: number } | null {
  if (!landmarks || landmarks.length < 33) return null;
  const mid = (a: { x: number; y: number }, b: { x: number; y: number }) => ({
    x: (a.x + b.x) / 2,
    y: (a.y + b.y) / 2,
  });
  const dist = (a: { x: number; y: number }, b: { x: number; y: number }) =>
    Math.hypot(a.x - b.x, a.y - b.y);

  const shoulderMid = mid(landmarks[SHOULDER_L], landmarks[SHOULDER_R]);
  const hipMid = mid(landmarks[HIP_L], landmarks[HIP_R]);
  const torsoLen = dist(shoulderMid, hipMid);
  if (!Number.isFinite(torsoLen) || torsoLen < 1e-9) return null;

  // MediaPipe's y axis grows downward, so image-down is (0, +1) and the
  // inclination is the angle of the torso vector against it.
  const cos = (hipMid.y - shoulderMid.y) / torsoLen;
  const torsoIncline = (Math.acos(Math.max(-1, Math.min(1, cos))) * 180) / Math.PI;

  const legLen =
    (dist(landmarks[ANKLE_L], landmarks[HIP_L]) +
      dist(landmarks[ANKLE_R], landmarks[HIP_R])) / 2;

  const ratio = legLen / torsoLen;
  if (!Number.isFinite(torsoIncline) || !Number.isFinite(ratio)) return null;
  return { torso_incline: torsoIncline, leg_torso_ratio: ratio };
}
